[RFC] Change return type to bool on buffer.c functions

war1025 opened the pull request:
Converts various functions to return `bool` that were already returning `TRUE` and `FALSE` only.

The return values should up-convert fine to `int` so none of the calling code should be affected.

war1025 committed:
Change return type to bool on buffer.c functions

Pyrohh commented:
This would be much easier to review if the style changes were kept to an absolute minimum.

Pyrohh commented on src/nvim/buffer.c:
```diff
@@ -3565,26 +3559,29 @@ void get_rel_pos(win_T *wp, char_u *buf, int buflen)
+  char_u *buf,
+  int buflen,
+  int add_file                   /* Add "file" before the arg number */
 )
```
e.g., I don't think lines such as 3565-3568 should be changed, especially since the arguments are still "old style" (not using `@param`, and comments aren't C++ style).

war1025 replied:
Fun fact: I looked at the style guide and they are actually supposed to be 4-spaced.

war1025 commented:
IMHO the style changes are kept to an absolute minimum. The return types on these functions need to be updated. There is no reason not to update the rest of the function signature to follow more closely the style guide.

Pyrohh commented:
> The return types on these functions need to be updated

Of course, I wasn't referring to that.

> There is no reason not to update the rest of the function signature to follow more closely the style guide.

It's much more code to review (minus the comments).

war1025 commented:
I just looked through the diff and I disagree that the function doc changes add an unreasonable overhead to the pull. 

Yes, there are changes that could strictly be left out. The code more closely conforms to the style guide because of these changes. I believe the benefit outweighs the added review time. Especially since this is a trivial set of changes.

Pyrohh commented:
Just to be clear, my first comment was only a suggestion, as I can't speak on behalf of all potential reviewers.  It's just that in my experience, a lot of the big and/or unwieldly PRs are left to rot because noone wants to review them (see test conversion PRS).

war1025 commented:
Fair enough. That has been my experience with most pulls I've put up, so I have to agree with you.

I was irritated earlier about something completely unrelated. I may have come across as overly aggressive with my comments. If so, I apologize. 

The thing that interests me about this project is that I have the opportunity to clean up the codebase of the editor I use every day. Neovim seems much more amenable to these cleanups than mainline Vim. Hopefully if I am patient and persistent enough I can make some headway :)

Pyrohh commented:
> I was irritated earlier about something completely unrelated. I may have come across as overly aggressive with my comments. If so, I apologize.

Oh, that's fine. I thought you were just being blunt, which can be very practical.

> Neovim seems much more amenable to these cleanups than mainline Vim. Hopefully if I am patient and persistent enough I can make some headway :)

Oh for sure, I think the style cleanups are helpful, and they are much appreciated.

One completely safe way to do style cleanups is to build your program, make a copy of it, do the style cleanup--making sure to only change text which wouldn't change the generated code (e.g., indentation/whitespace), then recompiling and comparing the new binary with the old (using `sha256`, for example). If you compile using the `-frandom-seed` GCC flag, and there are no other sources of non-determinism in the build ([see here](https://www.chromium.org/developers/testing/isolated-testing/deterministic-builds)), the binaries should be the same and you have a guarantee that there are no functional changes. I recall this being used a lot in LibreSSL, presumably because of how useful it is once everything is set up.

Unfortunately in Neovim, there are a few sources of non-determinism preventing such a setup, including absolute paths being encoded into the binary, as well as compilation times (`__DATE__` and `__TIME__`). If you're still interested, I found this useful:
http://blog.mindfab.net/2013/12/on-way-to-deterministic-binariy-gcc.html

That's all from me; if what I mention is actually wanted it should probably be broken off into a separate issue.

war1025 commented:
That's really interesting. I've glossed over some articles on the topic before, but I may have to delve more deeply into the subject now :)

bfredl commented:
I think it's fine (and even encouraged) to convert function comments to doxygen when the signature is changed (like int => bool return) but IMHO a function comment should then either be completely converted with `@param` and `@return` as @Pyrohh mentioned, or not converted at all. Leaving half-converted signature docs could actually be worse, because at a quick glance it may look as if the doc is (completely) converted, even though it isn't.

bfredl commented:
The unit tests are also failing, but it looks fine to just update to the new actual values (`true`/`false`).

war1025 commented:
@Pyrohh @bfredl I updated this pull to make all the functions I changed have doxygen style docs. They are not the greatest comments, but they are as good or better than what was there before.

I still need to update the tests to pass again

bfredl commented on src/nvim/edit.c:
```diff
@@ -7045,16 +7074,18 @@ static void ins_ctrl_hat(void)
+///
+/// @return true when leaving insert mode, false when going to repeat the
+///         insert.
+static bool ins_esc(
```
Signatures should be joined to single line (as long as it fits)

bfredl commented:
Just from some quick reading, there is no pattern of change I really object to, but the PR is quite big (999 changed lines) for a style PR. A way to get the changes reviewed and merged quicker, without requiring too much extra work, could be to split it filewise (I take it that changes in different files are really independent), not necessarily every file but at least the bigger ones like `edit.c` and `fileio.c`

war1025 commented:
@bfredl Done.
